File: beagle-declarative/src/main/kotlin/br/com/zup/beagle/widget/layout/FlexWidget.kt
package br.com.zup.beagle.widget.layout

import br.com.zup.beagle.widget.core.Appearance
import br.com.zup.beagle.widget.core.AppearanceWidget
import br.com.zup.beagle.widget.core.Flex
import br.com.zup.beagle.widget.core.Widget

data class FlexWidget(
    val flex: Flex? = null,
    val children: List<Widget>,
    override val appearance: Appearance? = null
) : Widget, AppearanceWidget

File: Sources/BeagleUI/Sources/Widgets/Layout/Flex/FlexWidget.swift
//
//  Copyright © 2019 Daniel Tes. All rights reserved.
//

import UIKit

public struct FlexWidget: Widget, HasAppearance {
    
    // MARK: - Public Properties
    
    public let appearance: Appearance?
    public let children: [Widget]
    public let flex: Flex
    
    // MARK: - Initialization
    
    public init(
        children: [Widget],
        flex: Flex = Flex(),
        appearance: Appearance? = nil
    ) {
        self.children = children
        self.flex = flex
        self.appearance = appearance
    }
    
   // MARK: - Configuration
    
    public func applyFlex(_ flex: Flex) -> FlexWidget {
        return FlexWidget(children: children, flex: flex)
    }
    
}

extension FlexWidget: Renderable {
    public func toView(context: BeagleContext, dependencies: Renderable.Dependencies) -> UIView {
        let containerView = UIView()
        
        children.forEach {
            let childView = $0.toView(context: context, dependencies: dependencies)
            containerView.addSubview(childView)
            dependencies.flex.enableYoga(true, for: childView)
        }
        containerView.applyAppearance(appearance)
        dependencies.flex.setupFlex(flex, for: containerView)
        
        return containerView
    }
}
